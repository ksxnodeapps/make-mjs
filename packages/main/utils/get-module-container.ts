import { iterateAncestorDirectories, joinPath } from '@make-mjs/utils'

export function * getModuleContainer (path: string, basename = 'node_modules') {
  for (const dir of iterateAncestorDirectories(path)) {
    yield joinPath(dir, basename)
  }
}

export default getModuleContainer
