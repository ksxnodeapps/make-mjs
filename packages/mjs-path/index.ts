export * from './lib/get-mjs-path'
export * from './lib/is-internal-module'
export * from './lib/is-mjs-path'
export * from './lib/from-file'
export * from './lib/from-dir'
export { default } from './lib/get-mjs-path'
