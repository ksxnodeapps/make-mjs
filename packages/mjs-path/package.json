{
  "name": "@make-mjs/mjs-path",
  "version": "0.0.0",
  "description": "Manipulate MJS module path",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/ksxnodeapps/make-mjs#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodeapps/make-mjs.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodeapps/make-mjs/issues"
  },
  "dependencies": {
    "exec-once": "^0.0.2",
    "@tsfun/object": "^0.0.1",
    "tslib": "^1.10.0",
    "@make-mjs/utils": "^0.0.0",
    "@make-mjs/fs-extra": "^0.0.0",
    "@types/node": "^12.7.5"
  }
}
