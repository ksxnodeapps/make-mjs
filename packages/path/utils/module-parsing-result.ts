export enum ModulePathKind {
  Internal = 'Internal',
  External = 'External'
}

export interface ModulePathParsingResultBase {
  readonly kind: ModulePathKind
  readonly name: string | null
  readonly path: string
}

export interface InternalModulePathParsingResult extends ModulePathParsingResultBase {
  readonly kind: ModulePathKind.Internal
  readonly name: null
}

export interface ExternalModulePathParsingResult extends ModulePathParsingResultBase {
  readonly kind: ModulePathKind.External
  readonly name: string
}

export type ModulePathParsingResult =
  InternalModulePathParsingResult |
  ExternalModulePathParsingResult
