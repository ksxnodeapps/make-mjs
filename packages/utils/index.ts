export type MaybePromise<X> = X | Promise<X>

export const REJECTION_SILENCER = (): void => undefined

/**
 * Silence a unhandled promise rejection warning
 * @param promise Promise to silence
 */
export function silenceRejection<X> (promise: Promise<X>) {
  void promise.catch(REJECTION_SILENCER)
  return promise
}

/**
 * Remove trailing `'/'` if there is one
 * @param path Path to be rid of trailing separator
 */
export function removeTrailingSeparator (path: string) {
  return path.endsWith('/')
    ? path.slice(0, -1)
    : path
}

/**
 * Like `path.join` but does not eliminate `'.'`
 * @param paths Paths to join together
 */
export function joinPath (...paths: string[]) {
  return paths
    .map(removeTrailingSeparator)
    .join('/')
}
