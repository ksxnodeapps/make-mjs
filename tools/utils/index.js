require('@tools/preloaded-node/register')
const lib = require('./lib')
module.exports = lib.esm(lib)
