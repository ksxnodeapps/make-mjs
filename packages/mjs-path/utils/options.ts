import { MaybePromise } from '@make-mjs/utils'

export type FullMjsPathOptions = Required<MjsPathOptions>

export interface ModulePathResolver {
  (options: FullMjsPathOptions): MaybePromise<string>
}

export interface ModulePathTester {
  (options: FullMjsPathOptions): MaybePromise<boolean>
}

export interface MjsPackageTesterOptions extends MjsPathOptions {
  readonly packageName: string
  readonly moduleEntry: string
}

export interface MjsPackageTester {
  (options: MjsPackageTesterOptions): MaybePromise<boolean>
}

export interface MjsPathOptions {
  readonly modulePath: string
  readonly moduleContainer: string
  readonly fromFile?: ModulePathResolver
  readonly fromDir?: ModulePathResolver
  readonly isMjsPath?: ModulePathTester
  readonly isInternalModule?: ModulePathTester
  readonly isMjsPackage?: MjsPackageTester
}

export interface FromDirOptions extends MjsPathOptions {
  readonly modulePath: string
  readonly moduleContainer: string
  readonly isMjsPackage: MjsPackageTester
  readonly newExt?: string
}
