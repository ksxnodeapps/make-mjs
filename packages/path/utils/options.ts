import { MaybePromise } from '@make-mjs/utils'
import { ModulePathParsingResult } from './module-parsing-result'

export type FullMjsPathOptions = Required<MjsPathOptions>
export type FullModulePathResolverOptions = Required<ModulePathResolverOptions>

export interface ModulePathResolver {
  (options: ModulePathResolverOptions): MaybePromise<string>
}

export interface ModulePathResolverOptions extends MjsPathOptions {
  readonly modulePath: string
  readonly moduleContainer: string
  readonly isMjsPackage: MjsPackageTester
  readonly modulePathParsingResult: ModulePathParsingResult
  readonly forceMjs: boolean
  readonly newExt?: string
}

export interface ModulePathTester {
  (options: FullMjsPathOptions): MaybePromise<boolean>
}

export interface MjsPackageTesterOptions extends MjsPathOptions {
  readonly packageName: string
}

export interface MjsPackageTester {
  (options: MjsPackageTesterOptions): MaybePromise<boolean>
}

export interface MjsPathOptions {
  readonly modulePath: string
  readonly moduleContainer: string
  readonly fromFile?: ModulePathResolver
  readonly fromDir?: ModulePathResolver
  readonly isMjsPath?: ModulePathTester
  readonly isInternalModule?: ModulePathTester
  readonly isMjsPackage?: MjsPackageTester
}
