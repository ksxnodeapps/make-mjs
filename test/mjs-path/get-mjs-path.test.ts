import { joinPath } from '@make-mjs/utils'
import { getMjsPath } from '@make-mjs/mjs-path'

import {
  MODULE_CONTAINER,
  INTERNAL_DIR_MODULE_PATH,
  INTERNAL_DIR_MODULE_PATH_SUPER,
  INTERNAL_FILE_MODULE_PATH,
  INTERNAL_FILE_MODULE_PATH_JS,
  INTERNAL_FILE_MODULE_PATH_MJS,
  INTERNAL_FILE_MODULE_PATH_SUPER,
  INTERNAL_FILE_MODULE_PATH_JS_SUPER,
  INTERNAL_FILE_MODULE_PATH_MJS_SUPER,
  INTERNAL_FILE_MODULE_PATH_NOT_EXIST,
  INTERNAL_FILE_MODULE_PATH_NOT_EXIST_JS,
  INTERNAL_FILE_MODULE_PATH_NOT_EXIST_MJS,
  INTERNAL_FILE_MODULE_PATH_NOT_EXIST_SUPER,
  INTERNAL_FILE_MODULE_PATH_NOT_EXIST_JS_SUPER,
  INTERNAL_FILE_MODULE_PATH_NOT_EXIST_MJS_SUPER,
  INTERNAL_NOT_FILE_NOR_DIR,
  EXTERNAL_MODULE_NAME,
  EXTERNAL_MODULE_NAME_MODULE,
  EXTERNAL_MODULE_NAME_BROWSER,
  EXTERNAL_MODULE_NAME_MAIN,
  EXTERNAL_MODULE_NAME_DEFAULT,
  pathExists,
  readJSON
} from './fsx-mocks'

const get = (modulePath: string) =>
  getMjsPath({ modulePath, moduleContainer: MODULE_CONTAINER })

describe('internal module', () => {
  describe('directory: append "/index.mjs"', () => {
    it('"./<name>"', async () => {
      expect(await get(INTERNAL_DIR_MODULE_PATH))
        .toBe(INTERNAL_DIR_MODULE_PATH + '/index.mjs')
    })

    it('"../<name>"', async () => {
      expect(await get(INTERNAL_DIR_MODULE_PATH_SUPER))
        .toBe(INTERNAL_DIR_MODULE_PATH_SUPER + '/index.mjs')
    })
  })

  describe('file: change extension to ".mjs"', () => {
    describe('"./<name>"', () => {
      it('without extension', async () => {
        expect(await get(INTERNAL_FILE_MODULE_PATH))
          .toBe(INTERNAL_FILE_MODULE_PATH + '.mjs')
      })

      it('with ".js" extension', async () => {
        expect(await get(INTERNAL_FILE_MODULE_PATH_JS))
          .toBe(INTERNAL_FILE_MODULE_PATH_JS.replace('.js', '.mjs'))
      })

      it('with ".mjs" extension', async () => {
        expect(await get(INTERNAL_FILE_MODULE_PATH_MJS))
          .toBe(INTERNAL_FILE_MODULE_PATH_MJS)
      })
    })

    describe('"../<name>"', () => {
      it('without extension', async () => {
        expect(await get(INTERNAL_FILE_MODULE_PATH_SUPER))
          .toBe(INTERNAL_FILE_MODULE_PATH_SUPER + '.mjs')
      })

      it('with ".js" extension', async () => {
        expect(await get(INTERNAL_FILE_MODULE_PATH_JS_SUPER))
          .toBe(INTERNAL_FILE_MODULE_PATH_JS_SUPER.replace('.js', '.mjs'))
      })

      it('with ".mjs" extension', async () => {
        expect(await get(INTERNAL_FILE_MODULE_PATH_MJS_SUPER))
          .toBe(INTERNAL_FILE_MODULE_PATH_MJS_SUPER)
      })
    })
  })

  describe('not exist: change extension to ".mjs"', () => {
    describe('"./<name>"', () => {
      it('without extension', async () => {
        expect(await get(INTERNAL_FILE_MODULE_PATH_NOT_EXIST))
          .toBe(INTERNAL_FILE_MODULE_PATH_NOT_EXIST + '.mjs')
      })

      it('with ".js" extension', async () => {
        expect(await get(INTERNAL_FILE_MODULE_PATH_NOT_EXIST_JS))
          .toBe(INTERNAL_FILE_MODULE_PATH_NOT_EXIST_JS.replace('.js', '.mjs'))
      })

      it('with ".mjs" extension', async () => {
        expect(await get(INTERNAL_FILE_MODULE_PATH_NOT_EXIST_MJS))
          .toBe(INTERNAL_FILE_MODULE_PATH_NOT_EXIST_MJS)
      })
    })

    describe('"../<name>"', () => {
      it('without extension', async () => {
        expect(await get(INTERNAL_FILE_MODULE_PATH_NOT_EXIST_SUPER))
          .toBe(INTERNAL_FILE_MODULE_PATH_NOT_EXIST_SUPER + '.mjs')
      })

      it('with ".js" extension', async () => {
        expect(await get(INTERNAL_FILE_MODULE_PATH_NOT_EXIST_JS_SUPER))
          .toBe(INTERNAL_FILE_MODULE_PATH_NOT_EXIST_JS_SUPER.replace('.js', '.mjs'))
      })

      it('with ".mjs" extension', async () => {
        expect(await get(INTERNAL_FILE_MODULE_PATH_NOT_EXIST_MJS_SUPER))
          .toBe(INTERNAL_FILE_MODULE_PATH_NOT_EXIST_MJS_SUPER)
      })
    })
  })
})

describe('external module', () => {
  describe('without "package.json"', () => {
    it('assertion: "package.json" does not exist', async () => {
      expect(await pathExists(
        joinPath(MODULE_CONTAINER, EXTERNAL_MODULE_NAME, 'package.json')
      )).toBe(false)
    })

    it('import by module name: append "/index.mjs"', async () => {
      expect(await get(EXTERNAL_MODULE_NAME))
        .toBe(EXTERNAL_MODULE_NAME + '/index.mjs')
    })

    it('import by file without extension: append ".mjs"', async () => {
      expect(await get(EXTERNAL_MODULE_NAME + '/target/file'))
        .toBe(EXTERNAL_MODULE_NAME + '/target/file.mjs')
    })

    it('import by ".js" file: change extension to ".mjs"', async () => {
      expect(await get(EXTERNAL_MODULE_NAME + '/target/file.js'))
        .toBe(EXTERNAL_MODULE_NAME + '/target/file.mjs')
    })

    it('import by ".mjs" file: does not change extension', async () => {
      expect(await get(EXTERNAL_MODULE_NAME + '/target/file.mjs'))
        .toBe(EXTERNAL_MODULE_NAME + '/target/file.mjs')
    })
  })

  describe('"package.json" has a "module" field', () => {
    const manifestContent = readJSON(
      joinPath(MODULE_CONTAINER, EXTERNAL_MODULE_NAME_MODULE, 'package.json')
    )

    it('assertion: "package.json" has a "module" field', async () => {
      expect(await manifestContent).toHaveProperty('module', expect.any(String))
    })

    it('assertion: "package.json" does not have "browser" field', async () => {
      expect(await manifestContent).not.toHaveProperty('browser')
    })

    it('assertion: "package.json" does not have "main" field', async () => {
      expect(await manifestContent).not.toHaveProperty('main')
    })

    it('import by module name: append entry name', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_MODULE))
        .toBe(EXTERNAL_MODULE_NAME_MODULE + '/' + (await manifestContent).module)
    })

    it('import by file without extension: append ".mjs"', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_MODULE + '/target/file'))
        .toBe(EXTERNAL_MODULE_NAME_MODULE + '/target/file.mjs')
    })

    it('import by ".js" file: change extension to ".mjs"', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_MODULE + '/target/file.js'))
        .toBe(EXTERNAL_MODULE_NAME_MODULE + '/target/file.mjs')
    })

    it('import by ".mjs" file: does not change extension', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_MODULE + '/target/file.mjs'))
        .toBe(EXTERNAL_MODULE_NAME_MODULE + '/target/file.mjs')
    })
  })

  describe('"package.json" has a "browser" field', () => {
    const manifestContent = readJSON(
      joinPath(MODULE_CONTAINER, EXTERNAL_MODULE_NAME_BROWSER, 'package.json')
    )

    it('assertion: "package.json" does not have "module" field', async () => {
      expect(await manifestContent).not.toHaveProperty('module')
    })

    it('assertion: "package.json" has a "browser" field', async () => {
      expect(await manifestContent).toHaveProperty('browser', expect.any(String))
    })

    it('assertion: "package.json" does not have "main" field', async () => {
      expect(await manifestContent).not.toHaveProperty('main')
    })

    it('import by module name: append entry name', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_BROWSER))
        .toBe(EXTERNAL_MODULE_NAME_BROWSER + '/' + (await manifestContent).browser)
    })

    it('import by file without extension: append ".mjs"', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_BROWSER + '/target/file'))
        .toBe(EXTERNAL_MODULE_NAME_BROWSER + '/target/file.mjs')
    })

    it('import by ".js" file: change extension to ".mjs"', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_BROWSER + '/target/file.js'))
        .toBe(EXTERNAL_MODULE_NAME_BROWSER + '/target/file.mjs')
    })

    it('import by ".mjs" file: does not change extension', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_BROWSER + '/target/file.mjs'))
        .toBe(EXTERNAL_MODULE_NAME_BROWSER + '/target/file.mjs')
    })
  })

  describe('"package.json" has a "main" field', () => {
    const manifestContent = readJSON(
      joinPath(MODULE_CONTAINER, EXTERNAL_MODULE_NAME_MAIN, 'package.json')
    )

    it('assertion: "package.json" does not have "module" field', async () => {
      expect(await manifestContent).not.toHaveProperty('module')
    })

    it('assertion: "package.json" does not have "browser" field', async () => {
      expect(await manifestContent).not.toHaveProperty('browser')
    })

    it('assertion: "package.json" has a "main" field', async () => {
      expect(await manifestContent).toHaveProperty('main', expect.any(String))
    })

    it('import by main name: append entry name', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_MAIN))
        .toBe(EXTERNAL_MODULE_NAME_MAIN + '/' + (await manifestContent).main!.replace('.js', '.mjs'))
    })

    it('import by file without extension: append ".mjs"', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_MAIN + '/target/file'))
        .toBe(EXTERNAL_MODULE_NAME_MAIN + '/target/file.mjs')
    })

    it('import by ".js" file: change extension to ".mjs"', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_MAIN + '/target/file.js'))
        .toBe(EXTERNAL_MODULE_NAME_MAIN + '/target/file.mjs')
    })

    it('import by ".mjs" file: does not change extension', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_MAIN + '/target/file.mjs'))
        .toBe(EXTERNAL_MODULE_NAME_MAIN + '/target/file.mjs')
    })
  })

  describe('"package.json" has neither "module" nor "browser" nor "main"', () => {
    const manifestContent = readJSON(
      joinPath(MODULE_CONTAINER, EXTERNAL_MODULE_NAME_DEFAULT, 'package.json')
    )

    it('assertion: "package.json" does not have "module" field', async () => {
      expect(await manifestContent).not.toHaveProperty('module')
    })

    it('assertion: "package.json" does not have "browser" field', async () => {
      expect(await manifestContent).not.toHaveProperty('browser')
    })

    it('assertion: "package.json" does not have "main" field', async () => {
      expect(await manifestContent).not.toHaveProperty('main')
    })

    it('import by main name: append "index.mjs"', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_DEFAULT))
        .toBe(EXTERNAL_MODULE_NAME_DEFAULT + '/index.mjs')
    })

    it('import by file without extension: append ".mjs"', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_DEFAULT + '/target/file'))
        .toBe(EXTERNAL_MODULE_NAME_DEFAULT + '/target/file.mjs')
    })

    it('import by ".js" file: change extension to ".mjs"', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_DEFAULT + '/target/file.js'))
        .toBe(EXTERNAL_MODULE_NAME_DEFAULT + '/target/file.mjs')
    })

    it('import by ".mjs" file: does not change extension', async () => {
      expect(await get(EXTERNAL_MODULE_NAME_DEFAULT + '/target/file.mjs'))
        .toBe(EXTERNAL_MODULE_NAME_DEFAULT + '/target/file.mjs')
    })
  })
})

it('when module is neither a file nor a directory: throw an error', async () => {
  await expect(get(INTERNAL_NOT_FILE_NOR_DIR)).rejects.toMatchSnapshot()
})
