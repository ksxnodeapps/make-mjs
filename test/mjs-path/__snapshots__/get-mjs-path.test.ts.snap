// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when module is neither a file nor a directory: throw an error 1`] = `[Error: Module is neither file nor directory: ./not-file-nor-dir]`;
